README.md
riak-dc
=======

So, Basho does not have an official javascript interface to Riak. However, there
is this [riak-js](http://riakjs.com/) thing. I found it to be poorly documented,
and the interface was not suitable for very small programs (I write command-line
tools, rather than web-front-end-ish type things). I wanted something
synchronous, something simple, that wouldn't take up too much space in my code.

Accordingly, I wrote a tiny little wrapper around `http`, and here is the product.

It is my hope that the code behind this never actually exceeds the length of the
documentation & test suite, and that anyone should be able to just
`npm install riak-dc` and be coding against a Riak quickly.

### Exported Functions

* `Riak.init( hostname, port, proxy )`
  - Takes two arguments: `hostname`, and `port`, which are what you would
  expect them to be. Note that this call is optional, and defaults are `localhost`
  and `8098`, respectively. If you wish to use an http proxy, you may pass an
  object that looks like `{ host: 'localhost', port: '3128' }` and `riak-dc`
  will attempt to use this proxy for all calls. You may also set the
  environment variables `HTTP_PROXY_HOST` and `HTTP_PROXY_PORT`. Using `init()`
  will override any environment variables you may have set.

* `Riak.ping()`
  - Takes no arguments, returns a true value if the server gives a 200 response,
  and something false otherwise.

* `Riak.get_keys( bucket )`
  - Takes a single argument, the name of the bucket you'd like keys
  for, and returns a promise to a list of the keys in that bucket.
  *Note that Riak does not care if this bucket does not exist.* If you ask for
  the keys from a non-existent bucket, Riak will dutifully tell you the list of
  keys in that bucket is `[]`, which is what `get_keys` will return to you here.

* `Riak.get_tuple( bucket, key )`
  - Takes two arguments, the name of the bucket and key you wish to
  have the value for. This is returned as a promised object, and will be whatever
  Riak has stored. Typically this is JSON, but you will need to parse that
  yourself (such as `JSON.parse(tuple)`), as it can be anything.
  *In the event that Riak has stored a 0-byte tuple:* (that is, the bucket/key pair
  are valid, the HTTP request returns 200, but the record is empty), you will
  receive an Error. This is not strictly an error condition, but it is helpful to
  know that the request itself was at least successful (whereas returning nothing
  is not helpful).

* `Riak.get_buckets( )`
  - Takes no arguments, and returns a promise of a list of the
  buckets Riak knows about.

* `Riak.put_tuple( bucket, tuple, key )`
<<<<<<< HEAD
  - Riak is very helpful in that it will provide you a key for the things you would
  store. So while in order to store something you must provide a bucket and the
  thing you would store (the tuple), the `key` argument is optional. If you supply
  a key, that will be provided as Riak for the tuple you have asked to store.
  User-supplied keys are useful for [key filters](http://docs.basho.com/riak/latest/dev/using/keyfilters/),
  if hierarchically-named keys is something you want.
  If you have omitted a key for storing the tuple, you will be returned a promise
  for the key that Riak has provided.

* `Riak.del_tuple( bucket, key )`
  - Should you wish to remove a tuple from your Riak, you must specify a bucket and
  a key. Returns what Riak returns; in the case you have tried to remove a tuple
  Riak doesn't know about, Riak will return an Error.

author
====

[@janearc](https://github.com/janearc), jane@cpan.org


Changelog.md
0.2.x
====

* Add ping() (#7)
* Expand unit tests to include 0-byte tuples & ping (#4)
* Ditched nopt-usage for sendak-usage in bin/riak-dc.js
* Add support for http proxy servers (#11)
* Add shorthand for riak-dc.js tool

0.1.x
====

* Initial revisions


test/riak-dc.js
var Riak   = require( '../lib/riak-dc.js' )
	, assert = require( 'assert' )
	, chai   = require( 'chai' )
	, cap    = require( 'chai-as-promised' )
	, nock   = require( 'nock' );

chai.use( cap );

it( 'Riak package initialised with test values', function () {
	assert.equal(Riak.init( 'localhost', '8098' ), 200)
} );

it( 'ping', function () {
	nock( 'http://localhost:8098', { 'Content-Type': 'application/json' } )
		.get( '/ping' )
		.reply( 200 );

	return Riak.ping().then( function (p) {
		assert( p, 'ping returns true' )
	} );
} );

it( 'get_buckets', function () {
	var buckets_rsp = { 'buckets': [ 'testing' ] };

	nock( 'http://localhost:8098', { 'Content-Type': 'application/json' } )
		.get( '/riak/?buckets=true' )
		.reply( 200, buckets_rsp );

	return Riak.get_buckets().then( function (buckets) {
		assert( buckets, 'defined return value' );
		assert.equal( buckets.length, 1, 'correct number of buckets' );
		assert.equal( buckets[0], 'testing', 'bucket 0 looks to be what we expected' );
	} );
} );

it( 'put_tuple (without key)', function () {
	var put_tuple_rsp = { 'json': 'true' };

	nock( 'http://localhost:8098'  )
		.post( '/riak/bucketname?returnbody=true' )
		.reply( 200, put_tuple_rsp, {
			'Content-Type': 'application/json',
			'Location': '/riak/bucketname/UQ47BRs5YSVepY2CEVrvuLY9EV7'
		} );
	return Riak.put_tuple( 'bucketname', put_tuple_rsp ).then( function ( key ) {
		assert( key, 'a key was returned' );
		assert.equal( key, 'UQ47BRs5YSVepY2CEVrvuLY9EV7', 'returned key is the key Riak supplied' );
	} );

} );

it( 'put_tuple (with key)', function () {
	var put_tuple_rsp = { 'json': 'true' };

	nock( 'http://localhost:8098' )
		.post( '/riak/bucketname/new_key_name?returnbody=true' )
		.reply( 200, put_tuple_rsp, {
			'Content-Type': 'application/json',
			'Location': '/riak/bucketname/new_key_name'
		} );

	return Riak.put_tuple( 'bucketname', put_tuple_rsp, 'new_key_name' ).then( function ( key ) {
		assert( key, 'a key was returned' );
		assert.equal( key, 'new_key_name', 'returned key is the key we supplied' );
	} );
} );

it( 'get_tuple (0-byte)', function () {
	nock( 'http://localhost:8098' )
		.get( '/riak/nodes/X0uzthV7wciJwNHjc2ymNqx4S5s' )
		.reply( 200, undefined, { 'Content-Type': 'application/json' } );

	return Riak.get_tuple( 'nodes', 'X0uzthV7wciJwNHjc2ymNqx4S5s' ).then( function ( tuple ) {
		if (tuple.constructor == Error) {
			// We have encountered a 0-byte tuple!
			assert( 1, 'returned error' )
		}
		else {
			assert( false, 'did not return an error' );
		}
	} );
} );

it( 'get_tuple', function () {
	var node_rsp = {
		'name':'first_node',
		'instance_id':'i-f7b8de1c',
		'availability_zone':'us-east-1a'
	};

	nock( 'http://localhost:8098' )
		.get( '/riak/nodes/X0uzthV7wciJwNHjc2ymNqx4S5s' )
		.reply( 200, node_rsp, { 'Content-Type': 'application/json' } );

	return Riak.get_tuple( 'nodes', 'X0uzthV7wciJwNHjc2ymNqx4S5s' ).then( function ( tuple ) {
		var parsed_tuple = JSON.parse(tuple);
		assert( tuple, 'tuple was returned' );
		assert( parsed_tuple, 'tuple parsed into json' );
		assert.deepEqual( parsed_tuple, node_rsp, 'returned value was not what was expected.' );
	} );

} );

it( 'get_keys', function () {
	// riak_long_reply{{{
	var riak_long_reply = {'props':{'name':'nodes','allow_mult':false,'basic_quorum':false,'big_vclock':50,'chash_keyfun':{'mod':'riak_core_util','fun':'chash_std_keyfun'},'dvv_enabled':false,'dw':'quorum','last_write_wins':false,'linkfun':{'mod':'riak_kv_wm_link_walker','fun':'mapreduce_linkfun'},'n_val':3,'notfound_ok':true,'old_vclock':86400,'postcommit':[],'pr':0,'precommit':[],'pw':0,'r':'quorum','rw':'quorum','small_vclock':50,'w':'quorum','young_vclock':20},'keys':['X0uzthV7wciJwNHjc2ymNqx4S5s']};

	// }}}

	nock( 'http://localhost:8098' )
		.get( '/riak/nodes?keys=true' )
		.reply( 200, riak_long_reply, { 'Content-Type': 'application/json' } );

	return Riak.get_keys( 'nodes' ).then( function (nodes) {
		assert( nodes, 'Riak returned something' );
		assert.deepEqual( nodes, [ 'X0uzthV7wciJwNHjc2ymNqx4S5s' ], 'list of nodes correct' );
	} );

} );

it( 'del_tuple', function () {
	nock( 'http://localhost:8098' )
		.delete( '/riak/nodes/X0uzthV7wciJwNHjc2ymNqx4S5s' )
		.reply( 204, '', { 'Content-Type': 'application/json' } );
	
	return Riak.del_tuple( 'nodes', 'X0uzthV7wciJwNHjc2ymNqx4S5s' ).then( function (r) {
		// So this should never really happen, but Riak doesn't really od much on
		// a successful delete.
		//
		if (r) {
			assert.notEqual( r, 'not found' );
		}
	} );
} );


examples/riak.js
#!/usr/bin/env node
// quick-and-dirty interface to riak from the shell
//

// parse opts
//
// var clean_args = require( 'components/common/js/supplemental.js' ).fix_quoted_array( process.argv );
var nopt = require('nopt')
	, noptUsage = require('nopt-usage')
	, Stream    = require('stream').Stream
	, path      = require('path')
	, knownOpts = {
			'list-buckets' : [ Boolean, null ],
			'list-keys'    : [ Boolean, null ],
			'get-tuple'    : [ Boolean, null ],
			'put-tuple'    : [ Boolean, null ],
			'del-tuple'    : [ Boolean, null ],
			'bucket'       : [ String, null ],
			'key'          : [ String, null ],
			'tuple'        : [ String ],
			'help'         : [ Boolean, null ]
		}
	, description = {
			'list-buckets' : 'Show all the buckets in riak',
			'list-keys'    : 'List all keys in a bucket',
			'get-tuple'    : 'Get a single tuple from a bucket/key pair',
			'put-tuple'    : 'Attempts to write a tuple to Riak; returns the serial Riak generates',
			'del-tuple'    : 'Attempts to delete a tuple front Riak; no return value',
			'bucket'       : 'To specify a bucket for operations',
			'key'          : 'To specify a key for operations',
			'tuple'        : 'To specify tuple for operations - this must be base64-encoded',
			'help'         : 'Sets the helpful bit.'
		}
	, defaults = {
			'help' : false
		}
	, shortHands = {
			'h'            : [ '--help' ],
		}
	, parsed = nopt(knownOpts, process.argv)
	, usage = noptUsage(knownOpts, shortHands, description, defaults)

if (parsed['help']) {
	// Be halpful
	//
	console.log( 'Usage: ' );
	console.log( usage );
	process.exit(0); // success
}

var Riak = require( 'components/common/js/riak-dc.js' ); // our riak synchronous wrapper

if (parsed['list-buckets']) {
	// Display the buckets in Riak
	//
	var pbuckets = Riak.get_buckets();
	pbuckets.then( console.log );
}

if (parsed['list-keys']) {
	// Display the keys in a given bucket
	//
	if (parsed['bucket']) {
		var bucket = parsed['bucket']
			, pkeys  = Riak.get_keys( bucket );
		pkeys.then( console.log )
	}
	else {
		console.log( 'You need to supply a bucket name if you want to list keys.' );
		console.log( usage );
		process.exit( -255 );
	}
}

if (parsed['get-tuple']) {
	// Display a given tuple (bucket/key pair)
	//
	if (parsed['bucket'] && parsed['key']) {
		var bucket = parsed['bucket']
			, key = parsed['key']
			, ptuple = Riak.get_tuple(bucket, key);

		ptuple.then( console.log );
	}
	else {
		console.log( 'You need to supply a bucket & key name if you want a tuple.' );
		console.log( usage );
		process.exit( -255 );
	}
}

if (parsed['put-tuple']) {
	// Place a tuple in Riak. Note that this needs to be a base64 encoded string.
	//
	if (parsed['bucket'] && parsed['tuple']) {
		if (parsed['key'] && !parsed['i-mean-it']) {
			console.log( 'Please let Riak decide the key to use for this tuple.' );
			console.log( usage );
			process.exit( -255 );
		}

		var third_arg = parsed['key'] ? parsed['key'] : undefined;

		var bucket = parsed['bucket']
			, tuple = new Buffer(parsed['tuple'], 'base64').toString('ascii')
			, presult = Riak.put_tuple(bucket, tuple, third_arg);

		console.log( 'attempted to place ' + tuple + ' in Riak' );

		presult.then( console.log );
	}
	else {
		console.log( 'You need to supply a bucket and data (a tuple) for this operation.' );
		console.log( usage );
		process.exit( -255 );
	}
}

// jane@cpan.org // vim:tw=80:ts=2:noet


lib/riak-dc.js
/*
 Riak is very fancy. Too fancy. This makes it much simpler. Following is
 a set of simple, synchronous-ish calls to get, put, and delete things
 from a Riak.

 riak don't care dot js
*/

'use strict';

var riak_host    = 'localhost'
	, riak_port    = 8098
	, proxy_port   = undefined
	, proxy_host   = undefined
	, transactions = { }
	, q            = require('q')
	, http         = require('http')

// chibi http client factory
//
var chcf = function ( path, method, headers, callback ) {
	var args = { };

	args['host']    = riak_host;
	args['port']    = riak_port;
	args['path']    = path;
	args['method']  = method;
	args['headers'] = headers;

	if ((process.env.HTTP_PROXY_PORT != undefined) &&
		(process.env.HTTP_PROXY_HOST != undefined)    ) {

		proxy_port = process.env.HTTP_PROXY_PORT;
		proxy_host = process.env.HTTP_PROXY_HOST;
	}

	if ( (proxy_port != undefined) && (proxy_host != undefined) ) {
		args['host']            = proxy_host;
		args['port']            = proxy_port;
		args['path']            = 'http://'.concat(riak_host, ':', riak_port, path);
		args['headers']['Host'] = riak_host;
		args['headers']['Port'] = riak_port;
	}

	return http.request( args, callback );
}
/**
 * Gets a list of all the buckets in riak, returned as a promise
 *
 */
function get_buckets () { // {{{
	// get_buckets returns a (promise of a) list of all the buckets riak knows about.
	//

	// We will store the response here later.
	//
	var gotten = '';

	var deferred = q.defer();

	var req = chcf( '/riak/?buckets=true', 'GET', {
		'Content-Type': 'application/json'
	}, function (result) {
		result.on('data', function (chunk) {
			// We assume this will be json that looks like:
			// { buckets: [ 'thing', ... ] }
			//
			var bucket_names = JSON.parse(chunk);
			deferred.resolve( bucket_names['buckets'] );
	} ) } );

	req.on( 'error', function(e) {
		return new Error("http: failure during request: " + e.message);
	} );

	[ 'connect', 'socket', 'end' ].forEach( function (state) {
		req.on( state, function () { default_logger( 'State ' + state + ' entered.' ) } )
	} );

	req.end();

	return deferred.promise;

} // }}} get_buckets

/**
 * Sends a ping to the database, asking for an HTTP 200, returns a promise
 * to an Error object if it fails.
 *
 */
function ping () { // {{{
	// ping returns a promise to a true value if your Riak replies in the
	// affirmative. If you get anything other than 200, your promise will be
	// false. At present there is no test for timeout.
	//
	var deferred = q.defer();

	var req = chcf( '/ping', 'GET', {
	}, function (response) {
		response.on('data', function (chunk) { deferred.resolve( true ) } );
		response.on( 'end', function () {      deferred.resolve( true ) } );
	} );

	[ 'connect', 'socket', 'end' ].forEach( function (state) {
		req.on( state, function () { default_logger( 'State ' + state + ' entered.' ) } )
	} );

	req.on( 'error', function(e) {
		deferred.resolve( false, new Error( 'Ping failed: ' + e ) );
	} );

	req.end();

	return deferred.promise;

} // }}}

/**
 * Gets a list of keys for a supplied bucket, returned as a promise.
 * @param {object} bucket - the name of the bucket
 */
function get_keys (bucket) { // {{{
	// get_tuples keys a (promise of a) list of all the keys that Riak knows
	// about in a given bucket.
	//

	// We will store the response here later.
	//
	var gotten = '';

	var deferred = q.defer();

	var path = '/riak/' + bucket + '?keys=true';

	var req = chcf( path, 'GET', {
		'Content-Type' : 'application/json'
	}, function (result) {
		result.on('data', function (chunk) {
			// We assume this will be json that looks like:
			// { keys: [ 'thing', ... ] }
			//
			var key_names = JSON.parse(chunk);
			deferred.resolve( key_names['keys'] );
	} ) } );

	req.on( 'error', function(e) {
		return new Error("http: failure during request: " + e.message);
	} );

	[ 'connect', 'socket', 'end' ].forEach( function (state) {
		req.on( state, function () { default_logger( 'State ' + state + ' entered.' ) } )
	} );

	req.end();

	return deferred.promise;

} // }}} get_keys

/**
 * Gets a tuple from Riak
 * @param {object} bucket - the name of the bucket
 * @param {object} key - the name of the key for the tuple you want.
 */
function get_tuple (bucket, key) { // {{{
	// get_tuple cannot return to you the stringy value of the bucket/key tuple,
	// because node. so instead we return to you a promise.
	//
	// TODO: Check for tombstones
	//

	var gotten = '';
	var subpath = bucket + '/' + key;

	var deferred = q.defer();

	var req = chcf( '/riak/'.concat( subpath ), 'GET', {
		'Content-Type' : 'application/json'
	}, function (response) {
		response.on('data', function (chunk) {
			gotten = gotten + chunk;
			deferred.resolve( gotten );
		} );
		response.on( 'end', function () {
			if (!gotten.length) {
				// So what has happened here is we have zero data from Riak, and we
				// have reached the end of our http response. This means we have a
				// zero-byte tuple. We have to let the user know about this, but it
				// is not strictly a bug.
				//
				deferred.resolve( new Error( '0-byte tuple encountered for key ' + key ) );
			}
		} );
	} );

	[ 'connect', 'socket', 'end' ].forEach( function (state) {
		req.on( state, function () { default_logger( 'State ' + state + ' entered.' ) } )
	} );

	req.on( 'error', function(e) {
		return new Error("http: failure during request: " + e.message);
	} );

	req.end();

	return deferred.promise;

} // }}} get_tuple

/**
 * Deletes a tuple from Riak (mostly)
 * @param {object} bucket - the name of the bucket
 * @param {object} key - the name of the key for the tuple you want.
 */
function del_tuple (bucket, key) { // {{{
	// del_tuple aims to delete a tuple given a bucket/key pair. The way Riak
	// works means that we don't /exactly/ delete it, but we mark it for later
	// deletion (a 'tombstone'). On success, Riak says nothing. If it's not
	// found, you'll get an Error.
	//

	var subpath = bucket + '/' + key;

	var deferred = q.defer();

	var req = chcf( '/riak/'.concat( subpath ), 'DELETE', {
		'Content-Type' : 'application/json'
	}, function (result) {
		result.on('data', function (chunk) {
			if (chunk == 'not found') {
				return new Error( 'tuple ' + bucket + '/' + key + ' not found' );
			}
			deferred.resolve( chunk );
	} ) } );

	[ 'connect', 'socket', 'end' ].forEach( function (state) {
		req.on( state, function () { default_logger( 'State ' + state + ' entered.' ) } )
	} );

	req.on( 'error', function(e) {
		return new Error("http: failure during request: " + e.message);
	} );
	req.end();

	return deferred.promise;
} // }}} del_tuple

/**
 * Places a tuple in a bucket in Riak.
 * @param {object} bucket - the name of the bucket
 * @param {object} payload - the object you wish to store
 * @param {object} key - if you wish to force the key name, supply it
 */
function put_tuple (bucket, payload, forced_key) { // {{{
	// put_tuple will send off your payload and do its best to return a promise
	// which contains a serial from the Riak you may use in the future to refer to it.
	//

	// Data from the request gets stored here, assuming chunking.
	//
	var gotten = ''
		, serial = '';

	var deferred = q.defer();

	if (forced_key) {
		bucket = bucket + '/' + forced_key;
	}

	var req = chcf( '/riak/'.concat( bucket, '?returnbody=true' ), 'POST', {
		'Content-Type' : 'application/json',
	}, function (result) {
		result.on('data', function (chunk) {
			gotten = gotten + chunk;
		} );
	} );

	req.on( 'error', function(e) {
		return new Error("http: failure during request: " + e.message);
	} );

	[ 'connect', 'socket', 'end' ].forEach( function (state) {
		req.on( state, function () { default_logger( 'State ' + state + ' entered.' ) } )
	} );

	// You must stringify things you post with the request object.
	//
	req.write( JSON.stringify(payload) );

	req.on( 'response', function ( response ) {

		// Riak returns us something like:
		//
		//   location: /riak/a_new_bucket/MYUbKWjuO5JvJGEoHHLI3ajfj5B
		//
		if (!forced_key) {
			var key;
			key = response.headers['location'].split('/')[3];

			// And hand it off to q.
			//
			deferred.resolve( key );
		}
		else {
			deferred.resolve( forced_key );
		}
	} );

	req.end();

	return deferred.promise;

} // }}} put_tuple

/**
 * Specify configuration options for the riak connection.
 * @param {string} host - the hostname/address of the Riak you wish to use
 * @param {string} port - the port to use on the supplied host
 * @param {object} proxy - a hash containing 'port' and 'host' referring to a proxy
 */
var init = function ( host, port, proxy ) { // {{{
	if (host) riak_host = host;
	if (port) riak_port = port;

	if (proxy) {
		proxy_port = proxy.port;
		proxy_host = proxy.host;
	}
	return 200;
} // }}} init

// Exported things
//

exports.init = init;

exports.get_keys    = get_keys;
exports.get_tuple   = get_tuple;
exports.get_buckets = get_buckets;
exports.put_tuple   = put_tuple;
exports.del_tuple   = del_tuple;
exports.ping        = ping;

// Not-exported things.
//
// TODO: Replace with log4js
//
function default_logger (s) {
	if (process.env['DEBUG']) {
		console.log(s);
	}
}

// @janearc // jane@cpan.org // vim:tw=79:ts=2:noet
